libs/header/sync/sync.go

```diff
@@ -256,49 +246,72 @@ func (s *Syncer[H]) doSync(ctx context.Context, fromHead, toHead H) (err error)
 
 // processHeaders gets and stores headers starting at the given 'from' height up to 'to' height -
 // [from:to]
-func (s *Syncer[H]) processHeaders(ctx context.Context, from, to uint64) (int, error) {
-	headers, err := s.findHeaders(ctx, from, to)
-	if err != nil {
-		return 0, err
-	}
-
-	amount, err := s.store.Append(ctx, headers...)
-	if err == nil && amount > 0 {
-		s.syncedHead.Store(&headers[amount-1])
+// processHeaders checks headers in pending cache that apply to the requested range.
+// If some headers are missing, it starts requesting them from the network.
+// All headers that will be received are verified to be contiguous.
+func (s *Syncer[H]) processHeaders(
+	ctx context.Context,
+	fromHead H,
+	to uint64,
+) (err error) {
+	cached := s.pending.cached(fromHead, to)
+	for _, cachedRange := range cached {
+		if fromHead.Height()+1 != cachedRange[0].Height() {
```

Comment:
are the `cached` ranges sorted? if not, we may have for instance:

- range `[10, 20]`
- range `[7, 25]`

and have `fromHead.Height() + 1` be 7, and thus on the first iteration end up doing an "unnecessary" external request to retrieve an already existing/cached height.

Reply:
ranges are sorted